src/vs/workbench/contrib/chat/browser/contrib/chatDynamicVariables.ts

```diff
@@ -60,60 +70,120 @@ export class ChatDynamicVariableModel extends Disposable implements IChatWidgetC
 							this.widget.refreshParsedInput();
 						}
 						return null;
-					} else if (Range.compareRangesUsingStarts(ref.range, c.range) > 0) {
+					}
+
+					if (Range.compareRangesUsingStarts(ref.range, c.range) > 0) {
 						const delta = c.text.length - c.rangeLength;
-						return {
-							...ref,
-							range: {
-								startLineNumber: ref.range.startLineNumber,
-								startColumn: ref.range.startColumn + delta,
-								endLineNumber: ref.range.endLineNumber,
-								endColumn: ref.range.endColumn + delta
-							}
+						ref.range = {
+							startLineNumber: ref.range.startLineNumber,
+							startColumn: ref.range.startColumn + delta,
+							endLineNumber: ref.range.endLineNumber,
+							endColumn: ref.range.endColumn + delta,
 						};
+
+						return ref;
 					}
 
 					return ref;
 				}));
 			});
 
-			this.updateDecorations();
+			this.updateVariableDecorations();
 		}));
 	}
 
-	getInputState(): any {
-		return this.variables;
-	}
-
 	setInputState(s: any): void {
 		if (!Array.isArray(s)) {
 			s = [];
 		}
 
 		this._variables = s;
-		this.updateDecorations();
+		this.updateVariableDecorations();
 	}
 
-	addReference(ref: IDynamicVariable): void {
-		this._variables.push(ref);
-		this.updateDecorations();
+	public addReference(ref: IDynamicVariable): void {
+		const variable = this._register(new ChatDynamicVariable(ref, this.fileService));
+
+		this._variables.push(variable);
+		this.updateVariableDecorations();
 		this.widget.refreshParsedInput();
+
+		// if the `prompt snippets` feature is enabled, start resolving
+		// nested file references immediatelly and subscribe to updates
+		if (ChatDynamicVariable.promptSnippetsEnabled(this.configService)) {
+			// subscribe to variable changes
+			this._register(variable.onUpdate(() => {
+				this.updateVariableTexts();
+				this.updateVariableDecorations();
+				this.widget.refreshParsedInput();
+			}));
+			// make sure the variable is updated on filesystem changes
+			variable.addFilesystemListeners();
+			// start resolving the file references
+			variable.resolve();
+		}
 	}
 
-	private updateDecorations(): void {
-		this.widget.inputEditor.setDecorationsByType('chat', dynamicVariableDecorationType, this._variables.map((r): IDecorationOptions => ({
-			range: r.range,
-			hoverMessage: this.getHoverForReference(r)
-		})));
+	/**
+	 * Update variables text inside input editor to add the `(+N more)`
+	 * suffix if the variable has nested child file references.
+	 */
+	private updateVariableTexts(): void {
+		for (const variable of this._variables) {
+			const text = `#file:${variable.filenameWithReferences}`;
```

Comment:
I liked the idea of having a distinct file extension for these files, which would also enable implementing attaching prompt snippets as a separate variable type instead of overloading the existing `#file:` variable and implicit context. I didn't see discussion of this route in the issue you posted but let me know if that was already evaluated as an option and dismissed. For context, we have `#kb:` for GitHub knowledge bases when using `@github`. I think that would result in a simpler and more contained implementation of this functionality, and also as a user I personally wouldn't expect to pick a prompt snippet through `#file`.

Reply:
Yeah we decided to revisit this after the first round of feedback, once we got some usage we should be in a better place to make this call 👍